### ISSUE TITLE
[Bug]: Removing colorbar misaligns other subplots

### ISSUE THREAD

**Issue by thoinka** [matplotlib/matplotlib#23502]
### Bug summary

When adding multiple colorbars with `plt.colorbar` to different subplots and then removing them again, the removal affects the width of the other subplot.

### Code for reproduction

```python
fig, ax = plt.subplots(2)
s = ax[0].scatter([0, 1, 2], [0, 1, 2], c=[0, 1, 2])
plt.colorbar(s, ax=ax[0])
cbar = plt.colorbar(s, ax=ax[1])
cbar.remove()
```

### Actual outcome

![mpl_bug_actual](https://user-images.githubusercontent.com/22074628/181497900-c58da451-fb68-414e-aef3-0e731e0933cf.png)

### Expected outcome

![mpl_bug_exp](https://user-images.githubusercontent.com/22074628/181497950-7db983b0-6f52-49b0-989f-523a2fc4f208.png)

### Additional information

This bug occurs with matplotlib>=3.4.0, matplotlib 3.3.4 produces the picture shown under "expected outcome", which is what I would assume is correct.

### Operating system

macOS 11.0.1 and Ubuntu 18.04

### Matplotlib Version

matplotlib 3.5.2

### Matplotlib Backend

module://ipykernel.pylab.backend_inline

### Python version

3.8.5

### Jupyter version

4.6.3

### Installation

pip

**Comment by oscargus** [comment]
Indeed! If anything, the bottom subplot should expand. Definitely not the top one.

**Comment by oscargus** [comment]
The problem seems to be that 
https://github.com/matplotlib/matplotlib/blob/1bac3b34d34bfd237571b08fe95404235ef7fc1f/lib/matplotlib/colorbar.py#L1087
points at the wrong axis, i.e. the top one. Which leads to that the rest of the code is applied to the top rather than the bottom axis.
https://github.com/matplotlib/matplotlib/blob/1bac3b34d34bfd237571b08fe95404235ef7fc1f/lib/matplotlib/colorbar.py#L1090-L1099

So `self.mappable.axes` is not set correctly(?).

**Comment by oscargus** [comment]
It is set correctly, it is "just" that the logic doesn't consider this case.

The code assumes that the mappable provided (the scatter plot) is also the object that got the colorbar attached to it and therefore the part that should be updated when the colorbar is removed. Probably one should use ax instead here. (Althiguh that may cause other issues...)

**Comment by jklymak** [comment]
Thanks @oscargus that does indeed seem to be the problem. I'm skeptical we should resize anything, but if that is the existing behaviour we should be storing the result of ax=ax and operating on that (if it is a single axes).  

The problem is that the colorbar axes is created outside of colorbar, so the colorbar method never knows if it stole from space anyone or not.  I think its overreach for colorbar to then try to give space back...

**Comment by jklymak** [comment]
Just to clarify here:

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots(2, layout='constrained')
s = ax[0].scatter([0, 1, 2], [0, 1, 2], c=[0, 1, 2])
plt.colorbar(s, ax=ax[0])

plt.show()
```

Does what I think the OP wants.  

If they really don't want to use constrained layout, they can also do various other hacks (inset_axes is the best way), or make a mock scalarmappable 

```python
fig, ax = plt.subplots(2)
s = ax[0].scatter([0, 1, 2], [0, 1, 2], c=[0, 1, 2])
plt.colorbar(s, ax=ax[0])
cbar = plt.colorbar(mpl.cm.ScalarMappable(norm=plt.Normalize(), cmap=plt.get_cmap('RdBu_r')),
ax=ax[1])
cbar.remove()
```

**Comment by timhoffm** [comment]
Duplicate #22257
